doc/devs.rst

```diff
@@ -0,0 +1,656 @@
+reliability.
+
+A channel is unaware of the underlying transport. Further, it leaves most
+details of the handshaking mechanism to the user as well. The class defines
+message types as well as interfaces for ``producer`` and ``consumer``
+implementations (both use CRTP to interface with user code).
+
+Channels in Data Store Actors
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In general, the master actor broadcasts state updates to its clones. This maps
+directly to the one-producer-many-consumers model of ``channel``. However,
+clones can also take the role a producer when forwarding mutating operations to
+the master.
+
+In a nutshell, the master actor (see ``master_actor.hh``) always has a producer
+attached to it and any number of consumers:
+
+.. code-block:: cpp
+
+  using producer_type = channel_type::producer<master_state>;
+
+  using consumer_type = channel_type::consumer<master_state>;
+
+  producer_type output;
+
+  std::unordered_map<entity_id, consumer_type> inputs;
+
+Conversely, the clone actor (see ``clone_actor.hh``) always has a consumer
+attached to it and it *may* have a producer:
+
+.. code-block:: cpp
+
+  using consumer_type = channel_type::consumer<clone_state>;
+
+  using producer_type = channel_type::producer<clone_state, producer_base>;
+
+  consumer_type input;
+
+  std::unique_ptr<producer_type> output_ptr;
+
+Clones initialize the field ``output_ptr`` lazily on the first mutating
+operation they need to forward to the master.
+
+Mapping to Command Messages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The message types defined in ``channel`` are never used for actor-to-actor
+communication directly. Instead, masters and clones exchange ``command_message``
+objects, which consist of a ``topic`` and an ``internal_command``. The essential
+interface for latter is defined as follows:
+
+.. code-block:: cpp
+
+  enum class command_tag {
+    action,
+    producer_control,
+    consumer_control,
+  };
+
+  class internal_command {
+  public:
+    // ...
+    using variant_type
+      = caf::variant<put_command, put_unique_command, put_unique_result_command,
+                     erase_command, expire_command, add_command, subtract_command,
+                     clear_command, attach_clone_command, attach_writer_command,
+                     keepalive_command, cumulative_ack_command, nack_command,
+                     ack_clone_command, retransmit_failed_command>;
+
+    detail::sequence_number_type seq;
+
+    entity_id sender;
+
+    variant_type content;
+  };
+
+  command_tag tag_of(const internal_command& cmd);
+
+Furthermore, data store actors define ``channel_type`` as
+``channel<entity_id, internal_command>``. When processing an
+``internal_command``, the receiver first looks at the tag.
+
+Control messages directly map to channel messages:
+
++-----------------------------------+-----------------------------------+
+| Internal Command Type             | Channel Message Type              |
```

Comment:
I don't quite follow the semantics of the channel messages. Are they hardcoded types that apply to all channels? If so, a summary of their semantics/use would be useful to have first.